python/lsst/daf/butler/datastores/fileDatastore.py

```diff
@@ -2490,9 +2491,17 @@ def transfer_from(
 
                     # Rely on source_records being a defaultdict.
                     source_records[missing].extend(dataset_records)
+            log.verbose("Completed scan for missing data files")
 
         # See if we already have these records
-        target_records = self._get_stored_records_associated_with_refs(refs, ignore_datastore_records=True)
+        if not dry_run:
+            target_records = self._get_stored_records_associated_with_refs(
+                refs, ignore_datastore_records=True
+            )
+        else:
+            # In dry-run mode we assume there is nothing in target repo.
+            # (otherwise everything would report that no work is needed.)
```

Comment:
I don't understand this.  In the typical production case, the file artifacts are mostly all present and the others are completely absent (since we run directly in the datastore root, so there are no file transfers), and the "work needed" is inserting stuff into the database.  What keeps the dry-run from doing the same, and then (doing the equivalent of) rolling back the DB transaction instead of committing it?

Reply:
Right, I've removed this. The loop now does act like a dry run in that if the dataset is known to the datastore already it doesn't even attempt to do the copy (the copy would never happen because that's masked out later with dry-run). I'm now wondering why this check doesn't happen at the start of the `Datastore.transfer_from()` on the basis that the target datastore thinking it has the files obviates the need to actually check the file system for them.